firewood/src/storage/linear/tests/mod.rs

```diff
@@ -67,3 +68,65 @@ fn reparent() {
     // proposal becomes Arc<LinearStore<Current<ConstBacked>>>
     // current becomes Arc<LinearStore<ConstBacked>>
 }
+
+#[derive(Debug)]
+pub struct InMemReadWriteLinearStore {
+    bytes: Vec<u8>,
+}
+
+impl InMemReadWriteLinearStore {
+    pub const fn new() -> Self {
+        Self { bytes: vec![] }
+    }
+}
+
+impl WriteLinearStore for InMemReadWriteLinearStore {
+    fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, std::io::Error> {
+        let offset = offset as usize;
+        if offset + object.len() > self.bytes.len() {
+            self.bytes.resize(offset + object.len(), 0);
+        }
+        self.bytes[offset..offset + object.len()].copy_from_slice(object);
+        Ok(object.len())
+    }
+}
+
+impl ReadLinearStore for InMemReadWriteLinearStore {
+    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read>, std::io::Error> {
+        let bytes = if addr as usize >= self.bytes.len() {
+            // Out of bounds. Return an empty cursor.
+            self.bytes[0..0].to_owned()
+        } else {
+            self.bytes[addr as usize..].to_owned()
+        };
```

Comment:
The warnings here are clues you can let the system do the checks for you:
```suggestion
        let bytes = self
            .bytes
            .get(addr as usize..)
            .unwrap_or_default()
            .to_owned();
```

Arguably the compiler would otherwise have to check the bounds twice (this might get optimized away?)